### PR TITLE
515 bump cxx std to 17

### LINKED ISSUE #515: Change used C++ standard to C++17
In order to allow new functionality, we should move from C++14 to C++17.

Currently, we only set C++14 for memilio general in https://github.com/DLR-SC/memilio/blob/main/cpp/memilio/CMakeLists.txt#L85

However, this should probably also be set explicitly in the models CMakeLists.

### PR DESCRIPTION
**Pull request by dabele**
## Changes

- set standard to c++17 in cmake
- fixed a few warnings (deprecations)
- upgrade pybind11 to fix a bug in cmake where pybind11 overwrites the c++ standard 

## Merge Request - GuideLine Checklist

- *Check our [git workflow](https://github.com/DLR-SC/memilio/wiki/git-workflow) before opening a Pull request/Merge request.*
- *Request a reviewer when your work is ready to review, before this please use the draft feature.*

### Checks by code author

- [x] Every addressed issue is linked (use the "Closes #ISSUE" keyword below)
- [x] New code adheres to [coding guidelines](https://github.com/DLR-SC/memilio/wiki/Coding-guidelines)
- [x] No large data files have been added (files should in sum not exceed 100 KB, avoid PDFs, Word docs, etc.)
- [x] Tests are added for new functionality and a local test run was successful
- [x] Appropriate **documentation** for new functionality has been added (Doxygen in the code and Markdown files if necessary)
- [x] Proper attention to licenses, especially no new third-party software with conflicting license has been added

### Checks by code reviewer(s)

- [x] Corresponding issue(s) is/are linked and addressed
- [x] Code is clean of development artifacts (no deactivated or commented code lines, no debugging printouts, etc.)
- [x] Appropriate **unit tests** have been added, CI passes and code coverage is acceptable (did not decrease)
- [x] No large data files added in the whole history of commits(files should in sum not exceed 100 KB, avoid PDFs, Word docs, etc.)

Closes #515

### CODE REVIEW

**Comment by dabele** [comment]
Because of the pybind upgrade, it may be necessary to delete your python build folder.

File: cpp/memilio/math/eigen.h [hunk 1]
```diff
@@ -24,6 +24,8 @@
 
 /* this file wraps includes from eigen3 library to disable warnings. */
 
+MSVC_WARNING_DISABLE_PUSH(4996)
```
Comment (mknaranja) [inline cpp/memilio/math/eigen.h:27]: pls provide a comment for future developers

**Review by HenrZu — COMMENTED** [review]
Looks good. The upgrade of the pybin version is also part of PR #626.
Had thought before that there might be compatibility issues there, but everything seems to work :)
I would approve it as soon as Martin's comment is solved.

File: cpp/memilio/math/eigen.h (hunk 1 above)
Comment (dabele) [reply cpp/memilio/math/eigen.h:28]: done

**Review by HenrZu — APPROVED** [review]

**Pull request merged**